File: java/amp-topology/global-topology-clients/java/src/main/java/amp/topology/client/GlobalTopologyService.java
package amp.topology.client;

import java.util.Map;
import java.util.concurrent.TimeUnit;

import amp.bus.rabbit.topology.ITopologyService;
import amp.bus.rabbit.topology.RoutingInfo;

import cmf.bus.EnvelopeHeaderConstants;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;

/**
 * An implementation of ITopologyService that utilizes a central
 * configuration and management system for topology, called
 * the "Global Topology Service".  This implementation relies on 
 * an external source to provide routing info.  If the routing info
 * is found, it is cached for some interval.  If not, the Service will
 * default to a "fallback" configuration.
 * 
 * @author Richard Clayton (Berico Technologies)
 */
public class GlobalTopologyService implements ITopologyService {

	public static long CACHE_EXPIRY_TIME_IN_SECONDS = 1000;
	
	Cache<String, RoutingInfo> routingInfoCache;
	
	IRoutingInfoRetriever routingInfoRetriever;
	
	FallbackRoutingInfoProvider fallbackProvider = null;
	
	public GlobalTopologyService(IRoutingInfoRetriever routingInfoRetriever){
		
		this(routingInfoRetriever, CACHE_EXPIRY_TIME_IN_SECONDS);
	}
	
	public GlobalTopologyService(
		IRoutingInfoRetriever routingInfoRetriever, 
		long cacheExpiryTime){
		
		this.routingInfoRetriever = routingInfoRetriever;
		
		this.routingInfoCache = 
			CacheBuilder
				.newBuilder()
				.expireAfterWrite(cacheExpiryTime, TimeUnit.SECONDS)
				.build();
	}
	
	public GlobalTopologyService(
		IRoutingInfoRetriever routingInfoRetriever, 
		long cacheExpiryTime, 
		FallbackRoutingInfoProvider fallbackProvider){
		
		this(routingInfoRetriever, cacheExpiryTime);
		this.fallbackProvider = fallbackProvider;
	}

	@Override
	public RoutingInfo getRoutingInfo(Map<String, String> routingHints) {
		
		String topic = routingHints.get(EnvelopeHeaderConstants.MESSAGE_TOPIC);
		
		RoutingInfo routingInfo = this.routingInfoCache.getIfPresent(topic);
		
		if (routingInfo == null){
			
			routingInfo = this.routingInfoRetriever.retrieveRoutingInfo(topic);
			
			if (routingInfo == null && this.fallbackProvider != null) {
				
				routingInfo = this.fallbackProvider.getFallbackRoute(topic);
			}
			
			// If the RoutingInfo is still null, end this program's life!
			if (routingInfo == null){
				
				throw new RoutingInfoNotFoundException(topic);
			}
			
			this.routingInfoCache.put(topic, routingInfo);
		}
		
		return routingInfo;
	}

	@Override
	public void dispose() {}
}

File: dotnet/source/services/amp.topology.client/GlobalTopologyService.cs
using System;
using System.Collections.Generic;
using System.Runtime.Caching;

using cmf.bus;

using amp.rabbit.topology;


namespace amp.topology.client
{
    /// <summary>
    /// An implementation of ITopologyService that utilizes a central
    /// configuration and management system for topology, called
    /// the "Global Topology Service".  This implementation relies on 
    /// an external source to provide routing info.  If the routing info
    /// is not found, it will default to a "fallback" configuration.
    /// </summary>
    /// <author>John Ruiz (Berico Technologies)</author>
    public class GlobalTopologyService : ITopologyService, IDisposable
    {
        private readonly IRoutingInfoRetreiver _routingRepo;
        private readonly IFallbackRoutingInfoProvider _fallbackProvider;


        public GlobalTopologyService(IRoutingInfoRetreiver routingRetreiver)
        {
            _routingRepo = routingRetreiver;
        }

        public GlobalTopologyService(
            IRoutingInfoRetreiver routingRetreiver,
            IFallbackRoutingInfoProvider fallbackProvider)
            : this(routingRetreiver)
        {
            _fallbackProvider = fallbackProvider;
        }


        public RoutingInfo GetRoutingInfo(IDictionary<string, string> routingHints)
        {
            // the routing we'll return
            RoutingInfo routing = null;


            // get the topic from the routing hints
            string topic = routingHints[EnvelopeHeaderConstants.MESSAGE_TOPIC];

            // guard clause -- make sure we have a topic
            if (null == topic) throw new ArgumentException("Cannot route a message with no topic");


            // try to get routing information from the injected info retriever
            routing = _routingRepo.RetrieveRoutingInfo(topic);

            // if we have no routing, ask the fallback provider
            if (routing.IsNullOrEmpty()) routing = _fallbackProvider.GetFallbackRoute(topic);

            // if even the fallback provider can't help us, we're done
            if (routing.IsNullOrEmpty()) throw new RoutingInfoNotFoundException("No routing for: " + topic);


            // finally, return the routing
            return routing;
        }

        public void Dispose()
        {
            // dispose the injected resources
            try { _routingRepo.Dispose(); }
            catch { }

            try { _fallbackProvider.Dispose(); }
            catch { }
        }
    }
}

